packages/calypso-e2e/src/rest-api-client.ts

```diff
@@ -0,0 +1,91 @@
+import fetch, { RequestInit } from 'node-fetch';
+
+type RequestMethods = 'post' | 'get';
+type RequestParams = {
+	method: 'post' | 'get';
+	headers: GenericDict;
+	body?: GenericDict;
+};
+type GenericDict = { [ key: string ]: string | number };
+
+interface Site {
+	ID: string;
+	name: string;
+	description: string;
+	URL: string;
+	site_owner: string;
+}
+interface MySitesResponse {
+	sites: Site[];
+}
+
+const BASE_URL = 'https://public-api.wordpress.com/rest/v1.1/';
+
+/**
+ * Client to interact with the WordPress.com REST API.
+ */
+export class RestAPIClient {
+	readonly bearerToken: string;
+
+	/**
+	 * Constructs an instance of the API client.
+	 */
+	constructor( bearerToken: string ) {
+		this.bearerToken = bearerToken;
+	}
+
+	/**
+	 * Builds and returns the request parameters.
+	 *
+	 * @param {RequestMethods} method Supported request methods.
+	 * @param {GenericDict} [body] Request body.
+	 * @returns {RequestParams} Completed request parameter.
+	 */
+	private buildRequestParams( method: RequestMethods, body?: GenericDict ): RequestParams {
+		const params: RequestParams = {
+			method: method,
+			headers: {
+				Authorization: `Bearer ${ this.bearerToken }`,
+				'Content-Type': 'application/json',
+			},
+		};
+
+		// POST requests require a body.
+		if ( method === 'post' ) {
+			if ( ! body ) {
+				throw new Error( 'Cannot build a POST request with undefined body.' );
+			}
+			params[ 'body' ] = body;
+		}
+
+		return params;
+	}
+
+	/**
+	 * Sends the request to the endpoint.
+	 *
+	 * @param {string} endpoint Endpoint to send the request.
+	 * @param {RequestParams} params Parameters for the request.
+	 */
+	private async sendRequest( endpoint: string, params: RequestParams ): Promise< any > {
+		// Trim the initial forward slash.
+		if ( endpoint.startsWith( '/' ) ) {
+			endpoint = endpoint.slice( 1 );
+		}
```

Comment:
Totally optional, but based on my experience on building similar API clients in the past, I'm not sure this is going to be the most useful abstraction over time.

Requests can vary a lot!  Some API requests may require specific headers for example. Also, POST requests may not necessarily be JSON -- they could Form or just URL-encoded search params. Accounting for all that variability in an abstraction like this over time can grow in a messy way sometimes. 

I've found in the past that what helps the most is to make really micro abstractions for small pieces (such as as a function to concat the host and route, or a function to get a "starting" header object with the auth token set). Then just have each public function (like `getMySites`) just build the request as needed. 

It makes each public function a little more verbose, but it makes it really obvious what that function is doing and how the web request is constructed.

Again, totally optional, but I wanted to throw out my perspective in case it's helpful! 😄

Reply:
Hmm. It's a lot to think about for sure.

I've found that the interactions with the WP API so far have followed a template of being JSON requests authorized using the bearer token. You're right that we might run into an odd call in the future that may be different. 

I'll look a bit more into what you said (to split things out into smaller builder methods).

Reply:
Yeah, it may be overkill, and what you have totally works!